### ISSUE TITLE
Healthchecking in Lua

### ISSUE THREAD

**Issue by canselcik** [envoyproxy/envoy#6901]
Hi,

I am trying to put deploy various custom and stateful origins behind an Envoy gateway. These origins may have arbitrary logic to their healthchecks such that for a certain type one may want to check if the response is 200 OK and valid JSON with a "foo" key and a "bar" value, and mark the origin healthy if and only if that's the case while some other origin might be okay with just a 200 OK and an empty response body.

As it stands, I think this kind of feature would involve creating a new Healthcheck Filter. Has this been abstracted away by another filter? Any suggestions for such a use case would be appreciated.

Thanks.

**Comment by mattklein123** [comment]
Not supported today. However, we support custom health checkers, so it should be possible to create a Lua one.

**Comment by canselcik** [comment]
Thanks @mattklein123. I will be experimenting with a healthcheck filter right now.

**Comment by canselcik** [comment]
@mattklein123 Sorry to mention you by name again but I am wondering if you could point me to a HealthcheckFilter that's "injected" into the Envoy build similarly to how the custom filters are "injected".

**Comment by mattklein123** [comment]
https://github.com/envoyproxy/envoy/tree/master/source/extensions/health_checkers/redis

**Comment by canselcik** [comment]
Thanks, that was the one I was looking at. There should be nothing preventing it to be built the way [envoyproxy/envoy-filter-example](https://github.com/envoyproxy/envoy-filter-example) is built, right?

**Comment by dio** [comment]
@canselcik do you mean to build it "out-of-tree"? I think so yeah. FWIW, I believe you have discovered that this custom health checker, if properly configured, will be loaded in here: https://github.com/envoyproxy/envoy/blob/003e2f417c4d58e99a8d15a26408a83be4e3e01c/source/common/upstream/health_checker_impl.cc#L71

**Comment by canselcik** [comment]
@dio Yes, thank you. Your answer actually pointed me to exactly where I should be looking at (which I had stumbled upon just earlier today), and I appreciate it.

FWIW, I have put together a minimal version and have been able to construct and get onInterval invoked on the healthchecker. It seems to be working without an issue.
